### ISSUE TITLE
Unexpected word token: "cascade"

### ISSUE THREAD

**Issue by unicornware** [oguimbal/pg-mem#192]
# Describe the bug

It seems that `CASCADE` is not implemented.

Because of this, I'm unable to use the library with [`sequelize@7.0.0-alpha.6`][1].

```
💔 Your query failed to parse.

This is most likely due to a SQL syntax error. However, you might also have hit a bug, or an unimplemented feature of pg-mem.
If this is the case, please file an issue at https://github.com/oguimbal/pg-mem along with a query that reproduces this syntax error.

👉 Failed query:

CREATE TABLE "user" ("id" SERIAL NOT NULL, "name" text NOT NULL, PRIMARY KEY ("id"));
DROP TABLE IF EXISTS "user" CASCADE;
;

💀 Syntax error at line 2 col 29:

  DROP TABLE IF EXISTS "user" CASCADE
                              ^
Unexpected word token: "cascade". Instead, I was expecting to see one of the following:

    - A "dot" token
    - A "semicolon" token
```

# To Reproduce

```sql
CREATE TABLE "user" ("id" SERIAL NOT NULL, "name" text NOT NULL, PRIMARY KEY ("id"));
DROP TABLE IF EXISTS "user" CASCADE;
```

check error on https://oguimbal.github.io/pg-mem-playground

# pg-mem version

`v2.3.3`

# Additional notes

- `sequelize.sync` options: `{ alter: { drop: false }, force: true }`
- the `sequelize` module is being used in my project via [`@nestjs/sequelize@8.0.0`][2]

[1]: https://github.com/sequelize/sequelize/tree/v7.0.0-alpha.6
[2]: https://github.com/nestjs/sequelize

**Comment by oguimbal** [comment]
Hi !

Will do.
In the meantime, as a workaround, you can skip this statement using [query interception](https://github.com/oguimbal/pg-mem#intercept-queries)

**Comment by unicornware** [comment]
@oguimbal thanks for the quick response and tip! i'll try it out and get back to you

**Comment by matt-sanders** [comment]
Having a similar issue here where I can't truncate a table which contains a foreign key referencing the same table. Try the following in the playground

```
CREATE TABLE public.users (
    user_id character varying(64) NOT NULL,
    parent_id character varying(64)
);

ALTER TABLE ONLY public.users
    ADD CONSTRAINT users_pkey PRIMARY KEY (user_id);

ALTER TABLE ONLY public.users
    ADD CONSTRAINT users_parent_id_fkey FOREIGN KEY (parent_id) REFERENCES public.users(user_id);

TRUNCATE TABLE users;
# OR
# TRUNCATE TABLE users CASCADE;
```
The `TRUNCATE TABLE users;` gives the following error:
```
ERROR: cannot truncate a table referenced in a foreign key constraint
DETAIL: Table "users" references "users".
HINT: HINT:  Truncate table "users" at the same time, or use TRUNCATE ... CASCADE.

🐜 This seems to be an execution error, which means that your request syntax seems okay,
but the resulting statement cannot be executed → Probably not a pg-mem error.

*️⃣ Reconsituted failed SQL statement: TRUNCATE TABLE users

👉 You can file an issue at https://github.com/oguimbal/pg-mem along with a way to reproduce this error (if you can), and  the stacktrace:
```

While `TRUNCATE TABLE users CASCADE;` gives:
```
💔 Your query failed to parse.
This is most likely due to a SQL syntax error. However, you might also have hit a bug, or an unimplemented feature of pg-mem.
If this is the case, please file an issue at https://github.com/oguimbal/pg-mem along with a query that reproduces this syntax error.

👉 Failed query:

    CREATE TABLE public.users (
    user_id character varying(64) NOT NULL,
    parent_id character varying(64)
);

ALTER TABLE ONLY public.users
    ADD CONSTRAINT users_pkey PRIMARY KEY (user_id);

ALTER TABLE ONLY public.users
    ADD CONSTRAINT users_parent_id_fkey FOREIGN KEY (parent_id) REFERENCES public.users(user_id);

TRUNCATE TABLE users CASCADE;;

💀 Syntax error at line 12 col 29:

  TRUNCATE TABLE users CASCADE;
                              ^
Unexpected semicolon token: ";". I did not expect any more input. Here is the state of my parse table:

    kw_continue → %word ● 
    kw_restart → %word ● 
```

Thanks!

**Comment by unicornware** [comment]
**update for those concerned: query interception worked just fine for this issue (and #193)**.

unfortunately, i ended up running into some other issues while trying to get the library integrated with my test suites, so i ended up moving away from `pg-mem`. afterwards i was able to get all my tests passing, but in hindsight, i'm not sure if the other errors were caused by how i was originally doing things, or the library itself (leaning more towards the former 😅).

---

for docker users interested in the workaround i landed on: 

i started by adding an initialization script to my [`postgres`][1] service to create multiple databases (rather than just one). 

next, i updated the `/etc/hosts` file on my local machine:

```hosts
# Docker Desktop
127.0.0.1 host.docker.internal
# End of section
```

and checked if `host.docker.internal` was reachable:

```shell
ping host.docker.internal
```

i use mocha to run tests and wanted to retain use of the [`growl`][2] feature. this solution gives me the ability to run test suites **outside of containers**,  thereby allowing me to get growls and use `host.docker.internal` as a database host.

in my test setup file, i added logic to configure, start, and stop a PostgreSQL server if `host.docker.internal` isn't reachable (requires PostgreSQL to be installed on the host (your local) machine, of course).

[1]: https://hub.docker.com/_/postgres
[2]: https://github.com/mochajs/mocha/wiki/Growl-Notifications

**Comment by oguimbal** [comment]
Hi,

I think that the docker solution is pretty well documented over the web, and the point of pg-mem is precisely to be able to avoid it (its complexity, and its execution overhead) for simple use cases.

Some details about the "other issues" you mentioned would be very much appreciated, instead of just advertizing a way to move away from pg-mem on the very repo of this library.
Which is kind of rude to its maintainer (which happens to be me), when you think about it.

Anyway, as for the "cascade" syntax error, yes, like your other issue, this is not implemented yet (but it will when I have some time, that's an easy feature to implement).

**Comment by unicornware** [comment]
@oguimbal 

i'm not advertising how to move away from your library, nor was i being rude. i posted **_my update_**, as i said i would. 

my update consisted of the fact that i believe "query interception worked just fine for this issue", but i can't be 100% in that because of the circumstances i described. i'm being intentionally vague about "other issues" because after hours of debugging and little tweaks, i simply don't remember all of the errors i encountered, nor is the list i was keeping up to date. on top of that, i more than clear when i stated i was not sure about the source of the errors, therefore there are no "details."

considering i read the project description, i know the point of pg-mem. an extremely large pet peeve of mine, however, is when developers post some type of update, mention a workaround, but don't make any mention of what the workaround is. i didn't do any research on the "docker solution" prior to, so i wouldn't have known if it was documented on the web or not. as i said, i was posting **_my update and the details i felt were relevant_.**

and finally, you're welcome 👍🏾

**Comment by oguimbal** [comment]
Fixed in `pg-mem@2.4.0`

**Comment by itamarmoa** [comment]
@oguimbal 

Hi,
I'm using v2.5 and having the error of cascade with truncate. Am I doing something wrong?

```
👉 Failed query:
    
        TRUNCATE TABLE table1, table2, table3 RESTART IDENTITY CASCADE;
    
    💀 Syntax error at line 1 col 333:
    
      TRUNCATE TABLE table1, table2, table3 RESTART IDENTITY CASCADE
                                                                                                                                                                                                                                                                                                                                                  ^
    Unexpected word token: "cascade". Instead, I was expecting to see one of the following:
    
        - A "semicolon" token

**Comment by oguimbal** [comment]
Hi @itamarmoa , the `cascade` flag on `truncate` statements was not supported.

This is fixed in `pg-mem@2.6.0`

**Comment by itamarmoa** [comment]
Thank you :)
